### ISSUE TITLE
Add count of reports to reduce duplication in payload

### ISSUE THREAD

**Issue by ScottHelme** [w3c/reporting#4]
In the new reporting API could we introduce a count for the number of times a report was triggered?

Right now the UA typically does de-duplication of reports to avoid sending multiple of the same POST requests. In the new reporting API I imagine a similar thing will happen and each payload will not contain duplicate reports. Whilst I understand the decision from a performance perspective, it could alter the perception of reports received when sites filter to see their 'most common report', for example. 

I'd suggest adding a count field the new report format:

    [{
      type: "csp",
      age: 10,
      url: "https://example.com/vulnerable-page/",
      count: 42
      report: {
        "blocked": "https://evil.com/evil.js",

This would allow the UA to report that this particular violation is occurring with a high frequency, but without consuming the traditionally large amount of resources required to do so.

**Comment by igrigorik** [comment]
> Right now the UA typically does de-duplication of reports to avoid sending multiple of the same POST requests.

Where / how is this done? 

De-duplication *is* nice, but it has all kinds of unrelated implications: what does `age` report exactly (newest, oldest, random report?), ditto for timestamp, a trivial change in payload will break this logic. In terms of consumed resources, compressing the request payload would get rid of most of the byte overhead.. if they're duplicates or close.

**Comment by ScottHelme** [comment]
I'm not sure on the specifics of how Chrome handles this but it does do de-duplication already with the existing report-uri mechanism. I believe it simply drops duplicate reports. Perhaps @mikewest could comment further?

You're right that it would impact the age value if we were to stack them like this and the total impact would depend on how long the browser cached reports for. Let's say the browser cached reports for a maximum of 60 seconds or until there were 100 in the cache as an example. I don't know what the current criteria are/would be but that would certainly allow us to catch multiple reports of the same violation. 

There's currently no mechanism in the existing functionality for the browser to indicate the time stamp of the violation in the report itself. On [report-uri.io](https://report-uri.io) I take the time stamp at the point I accept the report but there could have been various delays in it reaching the service. I think the main thing for reports like CSP is that we receive them and know the violation happened reasonably quickly, rather than knowing the precise second the violation happened. 

Compression in the HTTP request would alleviate most of the burden of the additional bytes on the wire, yes, but depending on the batch size it could still result in multiple HTTP requests being made where a single request would have been sufficient. If there are 50 duplicate reports and a batch size of 10 the browser would still need to do 5 POST requests. There's also additional burden on the endpoint receiving the reports as single units and whilst I agree it's small, when processing billions of reports per month, it adds up!

**Comment by igrigorik** [comment]
> On report-uri.io I take the time stamp at the point I accept the report but there could have been various delays in it reaching the service. I think the main thing for reports like CSP is that we receive them and know the violation happened reasonably quickly, rather than knowing the precise second the violation happened.

That's not an assumption we should rely on. For example, the user can be offline and trigger a set of exceptions -- it's important not to loose this information and relay it correctly to the collector.

**Comment by mikewest** [comment]
> I'm not sure on the specifics of how Chrome handles this but it does do de-duplication already with the existing report-uri mechanism.

Chrome holds a list of hashes of each report sent by a given document, and will skip POSTing a new report if it's an exact match for an old report: see the code around https://cs.chromium.org/chromium/src/third_party/WebKit/Source/core/frame/csp/ContentSecurityPolicy.cpp?rcl=5eeb08e6e66904d0ee2c96887f98caf8168f73d6&l=1272 for details.

@ckerschb would know if Firefox does something similar. Skimming WebKit's code, it doesn't look like they do.

**Comment by ScottHelme** [comment]
@igrigorik I see what you mean, but if the user is offline then isn't the time that the violation happened arbitrary? We don't know when the user went offline, just that they triggered the report offline. They could have been offline for minutes/hours/days which could impact the version of a document they have.

One way I could see this being helpful is if we had a problem, we patched, and a user came online 1 day later and sent their reports for it. It could be misleading to get them after the patch but we'd know they were triggered before. That seems a bit of a stretch for me, could you highlight some of the uses for the age value that you're thinking of?

**Comment by igrigorik** [comment]
The decision to discard-or-not should be up to the collector and the app in question.. Reporting is a generic mechanism that can be used by many different subsystems -- e.g. report navigation failures in offline mode due to routing mishaps, security violations, etc. In some cases the app may end up discarding such stale reports, but in others the reports may be just (if not more) as useful.

Stepping back, my main point here: don't assume that report generation time ~ delivery time, we should account for cases where the delta between those two can be large.

**Comment by ScottHelme** [comment]
"The decision to discard-or-not should be up to the collector and the app in question" - but Chrome already discards reports without any input from the host or operator of the reporting endpoint. I can see why, but if knowing how long ago the report was generated is important, is it not also just as important, or more important, to know how many occurrences there were? 

I don't know what the time window for de-duplication by Chrome is, but let's say the report fires every page load. Chrome would report the first event with it's appropriate time stamp but the repeated events would not be reported. @mikewest what clears the list of hashes for duplicate reports Chrome won't send? Are they applied per origin/URL/session?

If we keep the age value then I agree the reports don't stack but it seems like the UA should also stop removing what it deems to be unnecessary duplication when the frequency of the report could have an impact on triage/severity of an issue.

**Comment by igrigorik** [comment]
>  I can see why, but if knowing how long ago the report was generated is important, is it not also just as important, or more important, to know how many occurrences there were?

Sure.

> I don't know what the time window for de-duplication by Chrome is, but let's say the report fires every page load. Chrome would report the first event with it's appropriate time stamp but the repeated events would not be reported. @mikewest what clears the list of hashes for duplicate reports Chrome won't send? Are they applied per origin/URL/session?

Perhaps that's all well and good.. for CSP. Reporting API is not limited to CSP though, and I think we should spec and design it to have well defined semantics: given a report with a timestamp, deliver the report with age set to delta between when it was queued vs delivered; if the upstream producer wants to collapse multiple reports into a single report -- fine.

**Comment by ScottHelme** [comment]
@igrigorik agreed, but doesn't this mean that Chrome should stop de-duping? cc @mikewest
